File: apps/audit-scan-service/src/issues.ts
import {
  LIGHTHOUSE_ISSUE_MAPPINGS,
  type IssueMapping,
} from '@danieljoffe.com/shared-audit';

export interface ParsedIssue {
  category: IssueMapping['category'];
  severity: IssueMapping['severity'];
  title: string;
  description: string;
  impact: string;
  fix_difficulty: IssueMapping['fixDifficulty'];
  technical_detail: Record<string, unknown>;
}

export function parseIssues(
  lighthouseResult: Record<string, unknown>,
  axeResult: Record<string, unknown>
): ParsedIssue[] {
  const issues: ParsedIssue[] = [];

  // Parse Lighthouse audits
  const audits =
    (lighthouseResult['audits'] as Record<string, Record<string, unknown>>) ??
    {};

  for (const [auditId, mapping] of Object.entries(LIGHTHOUSE_ISSUE_MAPPINGS)) {
    const audit = audits[auditId];
    if (!audit) continue;

    const score = audit['score'] as number | null | undefined;
    if (score === null || score === undefined || score >= 0.9) continue;

    const displayValue = (audit['displayValue'] as string) ?? '';
    const numericValue = (audit['numericValue'] as number) ?? 0;

    const description = mapping.descriptionTemplate.replace(
      '{value}',
      displayValue
    );

    issues.push({
      category: mapping.category,
      severity: mapping.severity,
      title: mapping.title,
      description,
      impact: mapping.impactTemplate,
      fix_difficulty: mapping.fixDifficulty,
      technical_detail: {
        auditId,
        score,
        displayValue,
        numericValue,
      },
    });
  }

  // Parse axe-core violations
  const violations =
    (axeResult['violations'] as Array<Record<string, unknown>>) ?? [];

  for (const violation of violations) {
    const violationId = violation['id'] as string;
    const alreadyCaptured = issues.some(i =>
      i.title.toLowerCase().includes(violationId.replace(/-/g, ' '))
    );
    if (alreadyCaptured) continue;

    const nodes = violation['nodes'] as Array<unknown> | undefined;
    const nodeCount = nodes?.length ?? 0;
    const impact = violation['impact'] as string;

    issues.push({
      category: 'accessibility',
      severity:
        impact === 'critical' || impact === 'serious'
          ? 'critical'
          : impact === 'moderate'
            ? 'warning'
            : 'info',
      title: (violation['help'] as string) ?? violationId,
      description: `${violation['description'] as string}. Found ${nodeCount} instance${nodeCount !== 1 ? 's' : ''} on this page.`,
      impact:
        'Fixing this improves usability for people using assistive technology and may improve your search ranking.',
      fix_difficulty: 'easy',
      technical_detail: {
        axeRuleId: violationId,
        impact,
        nodeCount,
        helpUrl: violation['helpUrl'] as string,
      },
    });
  }

  // Sort: critical first, then warning, then info
  // Within each severity: performance -> accessibility -> seo -> ux
  const severityOrder = { critical: 0, warning: 1, info: 2 };
  const categoryOrder = { performance: 0, accessibility: 1, seo: 2, ux: 3 };

  issues.sort((a, b) => {
    const sevDiff = severityOrder[a.severity] - severityOrder[b.severity];
    if (sevDiff !== 0) return sevDiff;
    return categoryOrder[a.category] - categoryOrder[b.category];
  });

  return issues;
}

File: apps/audit-api/app/services/issues.py
from dataclasses import dataclass
from typing import Any

from app.services.issue_mappings import (
    LIGHTHOUSE_ISSUE_MAPPINGS,
    Category,
    FixDifficulty,
    Severity,
)

_SEVERITY_ORDER: dict[Severity, int] = {"critical": 0, "warning": 1, "info": 2}
_CATEGORY_ORDER: dict[Category, int] = {
    "performance": 0,
    "accessibility": 1,
    "seo": 2,
    "ux": 3,
}


@dataclass(frozen=True)
class ParsedIssue:
    category: Category
    severity: Severity
    title: str
    description: str
    impact: str
    fix_difficulty: FixDifficulty
    technical_detail: dict[str, Any]


def parse_issues(
    lighthouse_result: dict[str, Any], axe_result: dict[str, Any]
) -> list[ParsedIssue]:
    issues: list[ParsedIssue] = []

    audits = lighthouse_result.get("audits") or {}
    for audit_id, mapping in LIGHTHOUSE_ISSUE_MAPPINGS.items():
        audit = audits.get(audit_id)
        if not audit:
            continue

        score = audit.get("score")
        if score is None or score >= 0.9:
            continue

        display_value = audit.get("displayValue") or ""
        numeric_value = audit.get("numericValue") or 0

        description = mapping.description_template.replace("{value}", str(display_value))

        issues.append(
            ParsedIssue(
                category=mapping.category,
                severity=mapping.severity,
                title=mapping.title,
                description=description,
                impact=mapping.impact_template,
                fix_difficulty=mapping.fix_difficulty,
                technical_detail={
                    "auditId": audit_id,
                    "score": score,
                    "displayValue": display_value,
                    "numericValue": numeric_value,
                },
            )
        )

    violations = axe_result.get("violations") or []
    for violation in violations:
        violation_id = violation.get("id", "")
        spaced = violation_id.replace("-", " ").lower()
        already_captured = any(spaced in i.title.lower() for i in issues)
        if already_captured:
            continue

        nodes = violation.get("nodes") or []
        node_count = len(nodes)
        impact = violation.get("impact", "")

        if impact in ("critical", "serious"):
            severity: Severity = "critical"
        elif impact == "moderate":
            severity = "warning"
        else:
            severity = "info"

        plural = "s" if node_count != 1 else ""
        description = (
            f"{violation.get('description', '')}. "
            f"Found {node_count} instance{plural} on this page."
        )

        issues.append(
            ParsedIssue(
                category="accessibility",
                severity=severity,
                title=violation.get("help") or violation_id,
                description=description,
                impact=(
                    "Fixing this improves usability for people using assistive "
                    "technology and may improve your search ranking."
                ),
                fix_difficulty="easy",
                technical_detail={
                    "axeRuleId": violation_id,
                    "impact": impact,
                    "nodeCount": node_count,
                    "helpUrl": violation.get("helpUrl", ""),
                },
            )
        )

    issues.sort(
        key=lambda i: (_SEVERITY_ORDER[i.severity], _CATEGORY_ORDER[i.category])
    )
    return issues
